### ISSUE TITLE
floating attributes in module aliases with `-no-alias-deps` are ignored

### ISSUE THREAD

**Issue by nojb** [ocaml/ocaml#11867]
This note is meant to bring a shortcoming of `-no-alias-deps` to the attention of the rest of the developers, even if the beheaviour makes sense (as far as I understand module aliases). This flag, and module aliases, are used by Dune to implement wrapping of libraries and executables (see ocaml/dune#6820 for a related issue over at the Dune issue tracker).

Consider a file `foo.ml` with contents:
```
[@@@alert deprecated "Do not use this module."]
let x = 12
```
and a file `main.ml` with contents:
```
module F = Foo
let _ = Foo.x
let _ = F.x
```
Compile with:
```
$ ocamlc -c foo.ml
$ ocamlc -no-alias-deps -c main.ml
File "main.ml", line 2, characters 8-13:
2 | let _ = Foo.x
            ^^^^^
Alert deprecated: module Foo
Do not use this module.
```
and note that no alert is triggered for `F.x` which goes through the module alias.

**Comment by gasche** [comment]
Wee could implement a different semantics where a module alias (here `F`) "inherits" the attributes of the aliased module (here `Foo`). But it is not clear to me that it is the right approach in general. For example one could imagine code like the following:

```ocaml
module Badly_named = struct
  [@@@alert deprecated "use Correctly_named instead"]
  ...
end

module Correctly_named = Badly_named
```

**Comment by mobileink** [comment]
I was going to say I think this is the correct behavior, insofar as `-no-alias-deps` means exactly "do not look at (into) aliased dependencies."  So the aliasing equation in your example introduces a link-time but not a compile-time dependency of Main on Foo.cm[o|x]. At compile time Main depends only on the interface of Foo.  At least that's the mental model I've developed.

But then I went to the manual, which in this case is a model of opacity:

"When the compiler flag -no-alias-deps is enabled, type-level module aliases are also exploited to avoid introducing dependencies between compilation units. Namely, a module alias referring to a module inside another compilation unit does not introduce a link-time dependency on that compilation unit, as long as it is not dereferenced; it still introduces a compile-time dependency if the interface needs to be read, i.e. if the module is a submodule of the compilation unit, or if some type components are referred to. Additionally, accessing a module alias introduces a link-time dependency on the compilation unit containing the module referenced by the alias, rather than the compilation unit containing the alias."

src: [8 Type-level module aliases](https://v2.ocaml.org/manual/modulealias.html#s%3Amodule-alias)

If you can figure out what that means you get a gold star. Maybe the `F.x` in your example _should_ trigger the alert. I cannot tell from the manual. Does `F.x` count as "dereferencing" a "module alias referring to a module inside another compilation unit"? Does it mean "the interface needs to be read"?

**Comment by david-maison-TrustInSoft** [comment]
I'm not an expert on OCaml internals or attributes, but (as the original author of ocaml/dune#6820), I have been quite surprised to not see any alert from a floating attribute `deprecated` (which are tags as fatal errors in our builds).

With the following example:

```
(* foo.ml *)
[@@@alert deprecated "Do not use this module"]

let foo = ()
```

```
(* main.ml *)

module Correctly_named = Foo [@alert "-deprecated"]

module Badly_named = Foo [@alert "-deprecated"]
[@@alert deprecated "Use Correctly_named"]

module Badly_named2 = Foo

let () =
  Badly_named.foo;
  Correctly_named.foo;
  Foo.foo
```

`ocamlc foo.ml main.ml` correctly reports the three wrong usages (`Badly_named2 = Foo`, `Badly_named.foo` and `Foo.foo`) and forces users to deal with `Correctly_named`. That's not elegant, but (for a library) if the module `Foo` is not exported, then only the attribute on `Badly_named` should be required.

To come back to the problem about `-no-alias-deps` (where the `Badly_named2 = Foo` alert is not triggered with the above example). From the [Alerts](https://v2.ocaml.org/manual/alerts.html) section of the manual:

> And here is another example as a floating attribute on top of an “.mli” file (i.e. before any other non-attribute item) or on top of an “.ml” file without a corresponding interface file, so that any reference to that unit will trigger the alert:

From my understanding, `module Badly_named2 = Foo` is a reference to the `Foo` unit and should trigger the alarm.

Moreover, in the following example:

```
module F = struct let x = 1 end
[@@alert deprecated "Use Correctly_named"]

module C = F
```

The alarm is triggered even with `-no-alias-deps`. So why is there such a difference?

**Comment by ccasin** [comment]
The current situation seems unfortunate - it's common for floating alerts to be accidentally suppressed with a module alias.

The solution suggested by @gasche sounds very reasonable to me (tweaking the semantics of module aliases with respect to alerts).  I agree it could break existing code in the example given, but I suspect that's quite rare - do we know of any examples of this?

**Comment by gasche** [comment]
I realize that I am myself confused by floating alerts. In the toplevel (5.1.0), compare:

```ocaml
# module M = struct let x = 1 end [@@alert deprecated];;
module M : sig val x : int end
# M.x;;
Alert deprecated: module M

- : int = 1
# module N = struct [@@@alert deprecated] let y = 2 end ;;
module N : sig val y : int end
# N.y;;
- : int = 2
```

It does not seem to be the case that toplevel floating attributes are equivalent to structure-item-level attributes. Maybe toplevel floating attributes in *compilation units* are handled specially? I don't know what the logic is, and am not surprised that we fail to obtain consistent behavior from such a hodge-podge of hacks (including module aliases).

**Comment by gasche** [comment]
Possibly relevant: https://github.com/ocaml/RFCs/pull/26 .

**Comment by ccasin** [comment]
> I realize that I am myself confused by floating alerts. 

The current behavior of floating alert attributes inside non-file-level structures/signatures seems to be intentional, as it is documented in the [manual](https://v2.ocaml.org/manual/alerts.html#s%3Aalerts).  However, I agree it is confusing.  When implementing the recent rework of the misplaced attributes warning I initially got this wrong (by assuming floating alerts in other signatures are well-placed), and then when I fixed it found that it was fairly common for users to have made the same mistake in our codebase.  I think it would be nice to change the behavior.

**Comment by tmarti2** [comment]
This problem with floating attributes is still present in OCaml 5.3, are there any plan on fixing it ?
In Frama-C we are using dune to create our [kernel library](https://git.frama-c.com/pub/frama-c/-/blob/master/src/dune), and since it is wrapped by `dune` (we do not want to write a main lib file with ~900 module aliases + alerts) we lose alerts when we want to deprecate a whole module.

For now we're annotating every functions & co individually, which works, but it can be quite tedious in files with many exported things, and having working floating attributes in this case would really help a lot.

**Comment by nojb** [comment]
> This problem with floating attributes is still present in OCaml 5.3, are there any plan on fixing it ?

There are no concrete plans, because the desired semantics are not clear, as you will see from the discussion.

> we do not want to write a main lib file with ~900 module aliases + alerts

You could generate such a file programmatically, no?

**Comment by tmarti2** [comment]
There is a way to generate such a file which would preserve alerts ?

**Comment by Octachron** [comment]
Yes, since the root issue is with the wrapping module generated by dune. If you generate the module yourself, you can read the first attribute of each module in the library and repeat it in the alias definition.

**Comment by nojb** [comment]
> There is a way to generate such a file which would preserve alerts ?

Yes, you need to generate something of the form:
```
module N = N
[@@deprecated "Do not use this module."]
```

**Comment by Octachron** [comment]
One solution might be to add an attribute for inheriting alerts on module aliases:
```ocaml
module X = Y [@@inherit alert]
```
with the idea that it would be rather easy for build system generating module aliases to add this attribute without breaking backward compatibility, nor requiring them to read the floating attributes of all files.
